exp/controllers/rosaroleconfig_controller.go

```diff
@@ -0,0 +1,604 @@
+
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&expinfrav1.ROSARoleConfig{}).
+		WithOptions(options).
+		WithEventFilter(predicates.ResourceHasFilterLabel(mgr.GetScheme(), log.GetLogger(), r.WatchFilterValue)).
+		Complete(r)
+}
+
+// +kubebuilder:rbac:groups=infrastructure.cluster.x-k8s.io,resources=rosaroleconfigs,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=infrastructure.cluster.x-k8s.io,resources=rosaroleconfigs/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=infrastructure.cluster.x-k8s.io,resources=rosaroleconfigs/finalizers,verbs=update
+
+func (r *ROSARoleConfigReconciler) Reconcile(ctx context.Context, req ctrl.Request) (res ctrl.Result, reterr error) {
+	log := logger.FromContext(ctx)
+
+	roleConfig := &expinfrav1.ROSARoleConfig{}
+	if err := r.Get(ctx, req.NamespacedName, roleConfig); err != nil {
+		if apierrors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{Requeue: true}, nil
+	}
+
+	log = log.WithValues("cluster", klog.KObj(roleConfig))
+	scope, err := scope.NewRosaRoleConfigScope(scope.RosaRoleConfigScopeParams{
+		Client:         r.Client,
+		RosaRoleConfig: roleConfig,
+		ControllerName: "rosaroleconfig",
+		Endpoints:      r.Endpoints,
+		Logger:         log,
+	})
+
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to create rosaroleconfig scope: %w", err)
+	}
+
+	// Always close the scope and set summary condition
+	defer func() {
+		conditions.SetSummary(scope.RosaRoleConfig, conditions.WithConditions(expinfrav1.RosaRoleConfigReadyCondition), conditions.WithStepCounter())
+
+		if err := scope.Close(); err != nil {
+			reterr = errors.Join(reterr, err)
+		}
+	}()
+
+	ocm, err := r.NewOCMClient(ctx, scope)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to create OCM client: %w", err)
+	}
+
+	ocmClient, err := rosa.ConvertToRosaOcmClient(ocm)
+	if err != nil || ocmClient == nil {
+		return ctrl.Result{}, fmt.Errorf("failed to create OCM client: %w", err)
+	}
+
+	if !roleConfig.ObjectMeta.DeletionTimestamp.IsZero() {
+		conditions.MarkFalse(scope.RosaRoleConfig, expinfrav1.RosaRoleConfigReadyCondition, expinfrav1.RosaRoleConfigDeletionStarted, clusterv1.ConditionSeverityInfo, "Deletion of RosaRolesConfig started")
+		return ctrl.Result{}, r.reconcileDelete(scope, ocmClient)
+	}
+
+	if controllerutil.AddFinalizer(scope.RosaRoleConfig, expinfrav1.RosaRoleConfigFinalizer) {
+		if err := scope.PatchObject(); err != nil {
+			return ctrl.Result{}, err
+		}
+	}
+
+	if scope.RosaRoleConfig.Status.OIDCID == "" {
+		err = r.createOIDCConfig(roleConfig, scope, ocmClient)
+		if err != nil {
+			conditions.MarkFalse(scope.RosaRoleConfig, expinfrav1.RosaRoleConfigReadyCondition, expinfrav1.RosaRoleConfigReconciliationFailedReason, clusterv1.ConditionSeverityError, "Failed to create OIDC Config: %v", err)
+			return ctrl.Result{RequeueAfter: time.Second * 60}, fmt.Errorf("failed to OICD Config: %w", err)
```

Comment:
The above line (and all the other returns with non empty result and non-nil error) will produce the following log entry:
```
I0722 11:52:38.495821      15 controller.go:345] "Warning: Reconciler returned both a non-zero result and a non-nil error. The result will always be ignored if the error is non-nil and the non-nil error causes requeuing with exponential backoff. For more details, see: https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/reconcile#Reconciler" controller="rosaroleconfig" 
```